### PR TITLE
Reduce reliance on DataReaderMetaData in model and layer API

### PR DESCRIPTION
**Pull request by tbennun**

### CODE REVIEW

File: include/lbann/layers/transform/slice_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  data_type_layer<TensorDataType>::setup_dims();
+
+  // Setup the slice points if they are to be established by the data reader
+  // TODO: Move this responsibility to another component (input layer)
+  if (m_set_slice_points_from_data_reader) {
+    std::vector<size_t> slice_points;
+    std::string slice_point_method_name = "'get_slice_points_from_reader'";
+
+    LBANN_WARNING("slice_points_from_reader is deprecated and will be removed "
+                  "in a future version.");
```
Comment (tbennun) [inline include/lbann/layers/transform/slice_impl.hpp:47]: apart from refactoring to the impl file, these lines are the extent of the changes made.

File: src/layers/io/input_layer.cpp [hunk 2]
```diff
@@ -77,21 +77,23 @@ void input_layer<T, L, D>::write_specific_proto(lbann_data::Layer& proto) const
 }
 
 template <typename TensorDataType, data_layout T_layout, El::Device Dev>
-void input_layer<TensorDataType, T_layout, Dev>::setup_dims(
-  DataReaderMetaData& dr_metadata)
+void input_layer<TensorDataType, T_layout, Dev>::setup_dims()
 {
-  data_type_layer<TensorDataType>::setup_dims(dr_metadata);
+  data_coordinator& dc = get_trainer().get_data_coordinator();
+  const DataReaderMetaData& dr_metadata = dc.get_dr_metadata();
```
Comment (tbennun) [inline src/layers/io/input_layer.cpp:83]: A new input layer should avoid this altogether

Comment (bvanessen) [reply src/layers/io/input_layer.cpp:83]: @tbennun I need to go back and revisit why we plumbed this through in the first place.

File: src/optimizers/adam.cu [hunk 3]
```diff
@@ -131,7 +132,7 @@ void adam<TensorDataType>::step_compute_gpu(AbsDistMatrixType& values,
                                 m_eps,
                                 m_beta1,
                                 m_beta2,
-                                m_adamw_weight_decay * lr,
+                                adjusted_weight_decay,
```
Comment (bvanessen) [inline src/optimizers/adam.cu:135]: @tbennun Is this supposed to be in this PR?

File: src/optimizers/adam.cu [hunk 4]
```diff
@@ -149,7 +150,7 @@ void adam<TensorDataType>::step_compute_gpu(AbsDistMatrixType& values,
...
```
Comment (bvanessen) [inline src/optimizers/adam.cu:153]: @tbennun Same here.

File: src/layers/io/input_layer.cpp (hunk 2 above)
Comment (bvanessen) [reply src/layers/io/input_layer.cpp:83]: @tbennun to answer my own question, this was a small step to get rid of data readers everywhere.

**Review by bvanessen — COMMENTED** [review]
@tbennun I have a few questions that I noted for this PR.

File: src/optimizers/adam.cu (hunk 3 above)
Comment (tbennun) [reply src/optimizers/adam.cu:135]: Refactored to #2337

**Review by bvanessen — APPROVED** [review]

**Pull request merged**